flink-runtime/src/main/java/org/apache/flink/runtime/watermark/SourceWatermark.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.watermark;
+
+import java.io.Serializable;
+
+/**
+ * This represents the watermark for a single source partition.
+ */
+public class SourceWatermark implements Serializable {
```

Comment:
I wonder if it should be qualified as `SourceWatermark` vs. just `Watermark`? Perhaps there are use cases for exchanging watermarks across subtasks that don't necessarily belong to a source. One such example could be operators that perform asynchronous operations. Related, do we want to allow for an identifier for the watermark so that within an application multiple independent groupings could be formed?

Reply:
One problem is that there is already a `Watermark` class, but I agree with Thomas' comment. In the future, not all "sources" might be actual physical sources in the pipeline.

Reply:
Yeah, my intention was to keep this very focused on the exact use case at hand -- to provide simple state sharing for watermarks in the service of the source synchronization effort.  This is why the very specific naming and lack of additional features like namespaces, etc.

If we were to generalize this more it would be good to understand some other specific use cases -- and also to consider whether it's important to tackle that here or just go with the simplest interface we need for the task at hand.

@tweise @aljoscha If we do something more general what are you thinking?  Something more like a hash table or a collection of namespaced hashtables?  Would we need to make the key and value types generic, etc?  Would we want to then distribute the entire hashtable to every sub-task?

Reply:
I could imagine scenarios where different sources have different synchronization. That could be supported with a grouping mechanism for the tasks that participate in the watermark sync. The RPC would pass the group/namespace identifier as additional parameter and only get back the watermark for that (hash table would remain internal).

Reply:
Sounds good.  Will update shortly.